superset-frontend/spec/javascripts/explore/components/DateFilterControl/utils_spec.ts

```diff
@@ -0,0 +1,339 @@
+          sinceGrain: 'day',
+          sinceGrainValue: -7,
+          untilDatetime: '2021-01-27T00:00:00',
+          untilMode: 'specific',
+          untilGrain: 'day',
+          untilGrainValue: 7,
+          anchorMode: 'now',
+          anchorValue: 'now',
+        }),
+      ).toEqual(
+        'DATEADD(DATETIME("2021-01-27T00:00:00"), -7, day) : 2021-01-27T00:00:00',
+      );
+    });
+
+    it('7) relative : now', () => {
+      expect(
+        customTimeRangeEncode({
+          sinceDatetime: '2021-01-20T00:00:00',
+          sinceMode: 'relative',
+          sinceGrain: 'day',
+          sinceGrainValue: -7,
+          untilDatetime: '2021-01-27T00:00:00',
+          untilMode: 'now',
+          untilGrain: 'day',
+          untilGrainValue: 7,
+          anchorMode: 'now',
+          anchorValue: 'now',
+        }),
+      ).toEqual('DATEADD(DATETIME("now"), -7, day) : now');
+    });
+
+    it('8) relative : today', () => {
+      expect(
+        customTimeRangeEncode({
+          sinceDatetime: '2021-01-20T00:00:00',
+          sinceMode: 'relative',
+          sinceGrain: 'day',
+          sinceGrainValue: -7,
+          untilDatetime: '2021-01-27T00:00:00',
+          untilMode: 'today',
+          untilGrain: 'day',
+          untilGrainValue: 7,
+          anchorMode: 'now',
+          anchorValue: 'now',
+        }),
+      ).toEqual('DATEADD(DATETIME("today"), -7, day) : today');
+    });
+
+    it('9) relative : relative (now)', () => {
+      expect(
+        customTimeRangeEncode({
+          sinceDatetime: '2021-01-20T00:00:00',
+          sinceMode: 'relative',
+          sinceGrain: 'day',
+          sinceGrainValue: -7,
+          untilDatetime: '2021-01-27T00:00:00',
+          untilMode: 'relative',
+          untilGrain: 'day',
+          untilGrainValue: 7,
+          anchorMode: 'now',
+          anchorValue: 'now',
+        }),
+      ).toEqual(
+        'DATEADD(DATETIME("now"), -7, day) : DATEADD(DATETIME("now"), 7, day)',
+      );
+    });
+
+    it('10) relative : relative (date/time)', () => {
+      expect(
+        customTimeRangeEncode({
+          sinceDatetime: '2021-01-20T00:00:00',
+          sinceMode: 'relative',
+          sinceGrain: 'day',
+          sinceGrainValue: -7,
+          untilDatetime: '2021-01-27T00:00:00',
+          untilMode: 'relative',
+          untilGrain: 'day',
+          untilGrainValue: 7,
+          anchorMode: 'specific',
+          anchorValue: '2021-01-27T00:00:00',
+        }),
+      ).toEqual(
+        'DATEADD(DATETIME("2021-01-27T00:00:00"), -7, day) : DATEADD(DATETIME("2021-01-27T00:00:00"), 7, day)',
+      );
+    });
+  });
+
+  describe('customTimeRangeDecode', () => {
+    it('1) default value', () => {
+      expect(
+        customTimeRangeDecode(
+          'DATEADD(DATETIME("2021-01-27T00:00:00"), -7, day) : 2021-01-27T00:00:00',
+        ),
+      ).toEqual({
+        customRange: {
+          sinceDatetime: '2021-01-20T00:00:00',
```

Comment:
I'm not sure I fully understand the purpose of the `customTimeRangeDecode` function. It is supposed to apply the `dateadd` and `datetime` functions? However, [according to the code](https://github.com/apache/superset/blob/5d816a6e75a77c426b8ce3e5a3bf3a00043439b8/superset-frontend/src/explore/components/controls/DateFilterControl/utils.ts#L118), this `sinceDatetime` seems to always return  `7_DAYS_AGO` when the input format is `relative : specific`.

Reply:
![image](https://user-images.githubusercontent.com/2016594/106096334-4feb8a80-6170-11eb-9825-e2c58070ee04.png)

Reply:
Then shouldn't `sinceDatetime` be set to `2021-01-27T00:00:00` here, just to be more consistent with what is set in the input values?

Do you also do decode/encode when switching between non-advanced modes (i.e. modes that doesn't allow free text)? Not sure how feasible it is, but gut feeling tells me that it might be easier and cleaner to use he same config object internally for all frames and only do decode/encode when the time picker needs to communicate with the outside world.

Reply:
- `customTimeRangeEncode` is used only in the CustomFrame.
- `customTimeRangeDecode` is used in 2 places: CustomFrame and guessFrame function in DateFilterControl.

customTimeRangeDecode can determine whether an expression can be used in CustomFrame.

Reply:
`sinceDatetime` set to `2021-01-27T00:00:00` just to be more consistent in CustomFrame.

Reply:
Thanks for the explanation.